word2vec_decade_experiment/02_word2vec_decade_analysis.py

```diff
@@ -16,58 +16,147 @@
 
-    decade_df_dict[f"{str(year)}-{str(year+9)}"] = pd.DataFrame(
-        {"fp": fp, "tp": tp, "precision": precision, "recall": recall}
-    )
+    shared_tokens = sorted(list(shared_tokens))
 
-# ## Plot Performance
+# ## Calculate Global and Local Distances Between Time Periods
 
-year = 2001
+# ### Align Models via Orthogonal Procrustes
 
-g = (
-    p9.ggplot(decade_df_dict[f"{str(year)}-{str(year+9)}"], p9.aes(x="fp", y="tp"))
-    + p9.geom_point()
-    + p9.geom_line(p9.aes(x=[0, 1], y=[0, 1]), linetype="dashed")
-)
-print(g)
+if not Path(aligned_model_file_path).exists():
+    years_analyzed = sorted(list(word_model_dict.keys()), reverse=True)
+    latest_year = years_analyzed[0]
+    aligned_models = {}
+
+    # Years must be in sorted descending order
+    for year_pair in window(years_analyzed):
+
+        year_pair_label = "_".join(map(str, year_pair[::-1]))
+
+        if year_pair[0] == latest_year:
+            aligned_models[str(year_pair[0])] = word_model_dict[year_pair[0]].wv[
+                shared_tokens
+            ]
+
+        else:
+
+            aligned_models[str(year_pair[0])] = word_model_dict[year_pair[0]].wv[
+                shared_tokens
+            ]
+
+            translation_matrix, scale = orthogonal_procrustes(
+                word_model_dict[year_pair[0]].wv[shared_tokens],
+                word_model_dict[latest_year].wv[shared_tokens],
+            )
 
-g = (
-    p9.ggplot(
-        decade_df_dict[f"{str(year)}-{str(year+9)}"],
-        p9.aes(x="recall", y="precision"),
+        translation_matrix, scale = orthogonal_procrustes(
+            word_model_dict[year_pair[1]].wv[shared_tokens],
+            word_model_dict[latest_year].wv[shared_tokens],
+        )
+
+        aligned_models[str(year_pair[1])] = (
+            word_model_dict[year_pair[1]].wv[shared_tokens] @ translation_matrix
+        )
+
+    aligned_models["shared_tokens"] = shared_tokens
+
+if not Path(aligned_model_file_path).exists():
+    pickle.dump(aligned_models, open(aligned_model_file_path, "wb"))
+
+# ### Calculate the Global and Local Distances between Words
+
+aligned_models = pickle.load(open(aligned_model_file_path, "rb"))
+years_analyzed = sorted(list(aligned_models.keys()), reverse=True)[1:]
+n_neighbors = 25
+
+shared_tokens = sorted(aligned_models["shared_tokens"])
+for key in tqdm.tqdm(window(years_analyzed)):
+
+    global_distance = get_global_distance(
+        aligned_models[key[0]], aligned_models[key[1]], aligned_models["shared_tokens"]
+    )
+
+    local_distance = get_local_distance(
+        aligned_models[key[0]],
+        aligned_models[key[1]],
+        aligned_models["shared_tokens"],
+        neighbors=n_neighbors,
+    )
+
+    label = "_".join(reversed(key))
+
+    (
+        global_distance.merge(local_distance)
+        .assign(shift=lambda x: x.global_dist.values - x.local_dist.values)
+        .to_csv(
+            f"output/{year_distance_folder}/{label}_dist.tsv", sep="\t", index=False
```

Comment:
Since you are using Path, maybe you can create this one with that class instead of using a string.